dataclients/kubernetes/crd.md

```diff
@@ -0,0 +1,601 @@
+
+Date: 2019-09-04
+Status: open
+Owners: @aryszka @szuecs
+
+This proposal will be kept up to date, if you like to discuss it
+please use https://github.com/zalando/skipper/issues/660
+
+Disclaimer:
+Skipper's `kubernetes` dataclient will also support Ingress in the
+future.
+
+## Purpose
+
+The purpose of the document is to define a CRD, that solves issues
+we observed, while using Ingress.
+
+There are some other solutions out there, like Istio, SMI and others,
+but all of these have drawbacks for skipper provided use cases. One
+common example is an application that needs a redirect to somewhere
+else for a specific path in an Ingress type of object. This example is
+only provided with multiple objects or annotations, both are unsafe
+for concurrent use.
+
+## Problem statement
+
+Provide a format for routing input that makes it easier to benefit
+more from Skipper's routing features, without workarounds, than the
+generic Ingress specs. Primarily targeting more complex Kubernetes
+Ingress scenarios.
+
+We want to reduce boilerplate, enable traffic switching, enable common
+use cases and enable API gateway features in a more safe way.
+
+## Background Information
+
+A RouteGroup is used to store dependent group of routes in one Object,
+that can act as one thing.
+Users of this CRD are multiple controllers and end-users.
+
+### Users
+
+#### Controllers
+
+For example [stackset-controller](https://github.com/zalando-incubator/stackset-controller)
+orchestrates traffic switching. In this case a set of routes with different
+configurations to one backend should be updated atomically.
+
+Another example is an openapi-spec controller, which goal is to create
+a set of routes based on API endpoint definitions with rules, defaults
+and overrides.
+
+#### End User
+
+An end-user might want to create an Ingress with an additional path
+related to the Ingress, that would do a redirect.
+
+### Goals of the RouteGroup CRD
+
+- more DRY object than ingress (hosts and backends separated from path rules)
+- support common use cases in a safe and DRY way
+- avoid defining unrelated route groups in the same specification object
+- better support for skipper specific features ([Predicates](https://opensource.zalando.com/skipper/reference/predicates) and [Filters](https://opensource.zalando.com/skipper/reference/filters/))
+- orchestrate traffic switching for a full set of routes (RouteGroup) without redundant configuration
+- enable complex cases: for example an enhanced openapi-spec controller write this object, instead of creating multiple ingress and [stackset-controller](https://github.com/zalando-incubator/stackset-controller) does traffic switching
+
+## Proposed solution
+
+We create a CRD with `kind: RouteGroup`, that can express the mentioned use cases.
+The spec of the RouteGroup has 3 keys that are all arrays of some kind.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+spec:
+  hosts:
+  - <string>
+  backends:
+  - <backend>
+  paths:
```

Comment:
@aermakov-zalando suggested `routes`, instead of `paths`. 
I could also imagine `routes_template`, but I think `routes` sounds more nice. Template, because dependent on number of `hosts`, the routes will be multiplied.

Reply:
changed `paths` to `routes`